Make `ProductHint`s and `CoproductHint`s more general

jcazevedo opened the pull request:
This PR proposes a refactoring to `ProductHint`s and `CoproductHint`s in order to make them more general. The idea behind this is to allow better customization of derived readers and writers, regardless of the derivation code used to produce them. Incidentally, this moves a lot of repeated code between shapeless-based and magnolia-based derivation to the hints. This also enables #356 (which I leave for a future PR).

This is obviously not binary compatible with older versions of PureConfig. However, only custom implementations of `ProductHint` (not created via `ProductHint.apply`) and `CoproductHint` need to be migrated.

This also modifies the derivation of readers for case classes in `pureconfig-magnolia` to accumulate errors, at the expense of using the unsafe `rawConstruct`. I ended up doing so because the `constructMonadic` interface of Magnolia doesn't play very well with the proposed `ProductHint` interface of returning both the result of reading a field and the next `ConfigObjectCursor` to use.

jcazevedo committed:
Make `ProductHint` a more general abstraction for customizing derivations

Allow `ProductHint` to encode the behavior of reading and writing values of
specific types from/to a `ConfigObject`.

jcazevedo committed:
Add a `sucess` method to `Result`

jcazevedo committed:
Improve `ProductHint` documentation

jcazevedo committed:
Make `CoproductHint` a more general abstraction for customizing derivations

Add better decision points to `CoproductHint`, in order to allow composing
errors and make better decisions on when to read next coproduct options.

jcazevedo committed:
Minor style fixes

jcazevedo committed:
Remove unused method

jcazevedo committed:
Remove difference in behavior from `pureconfig-magnolia`s README

jcazevedo committed:
Make `CoproductHint` compile against Scala 2.11

jcazevedo committed:
Fix tut output

ruippeixotog commented on modules/generic-base/src/main/scala/pureconfig/generic/CoproductHint.scala:
```diff
@@ -57,6 +36,19 @@ trait CoproductHint[T] {
+   * @return the config for the sealed family or coproduct wrapped in a `Right`, or a `Left` with the failure if some
+   *         error occurred.
+   */
+  def to[A <: T](writer: ConfigWriter[A], name: String, value: A): ConfigReader.Result[ConfigValue]
```
Same thing here regarding `writer`: what are the advantages of providing `writer` and `value` here? All our three hint types call `writer.to(value)` and it seems natural to do so - if other behavior is needed it would probably be more elegant for users provided a custom writer for `A` instead.

ruippeixotog replied:
I know this was like this before, but why do we return a `ConfigReader.Result` here if the write path does not handle errors? Since we are changing the API, shouldn't we drop `Result` here and throw exceptions on hints if needed? As it is it may be misleading to users.

jcazevedo replied:
> I know this was like this before, but why do we return a `ConfigReader.Result` here if the write path does not handle errors? Since we are changing the API, shouldn't we drop `Result` here and throw exceptions on hints if needed? As it is it may be misleading to users.

I dropped the `Result` and started throwing `CoproductHintException`s in 977aacf323b3d687bd458943595a01d97c8cc7b9. A `CoproductHintException` wraps a single `FailureReason`, which is enough for our failure scenarios in hints.

ruippeixotog commented on modules/magnolia/src/main/scala/pureconfig/module/magnolia/MagnoliaConfigReader.scala:
```diff
@@ -67,27 +52,15 @@ object MagnoliaConfigReader {
+    def from(cur: ConfigCursor): ConfigReader.Result[A] = {
+      ctx.subtypes.foldRight[ConfigReader.Result[A]](Left(hint.noOptionFound(cur))) {
+        case (subtype, rest) =>
+          hint.from(cur, subtype.typeclass, subtype.typeName.short, rest)
```
Even though `rest` is lazy this implementation does not make use of it, which means that `hint.from` will always be called with all subtypes regardless of what the hint does.

ruippeixotog commented on modules/generic-base/src/main/scala/pureconfig/generic/CoproductHint.scala:
```diff
@@ -14,39 +14,18 @@ import pureconfig.syntax._
-   * @return `true` if the next class or coproduct option should be tried, `false` otherwise.
-   */
-  def tryNextOnFail(name: String): Boolean
+  def from[A <: T](cur: ConfigCursor, reader: ConfigReader[A], name: String, rest: => ConfigReader.Result[T]): ConfigReader.Result[T]
```
I don't like this new API very much, for two reasons:
- We're changing this to a kind of continuation-passing style, which seems like a clumsy interface for users (why do you need to use specifically that value at the end to fall back?) besides allowing more than one possible implementation for the same behavior (can we do `rest.left.flatMap(_ => /* try reading this subtype */)` instead of the canonical way? Which one is more efficient?). I always thought of derivation code as owning the control flow, the "algorithm" to try the different options.
It can prevent some duplicated logic in both derivation libs, but it did it at the cost of adding extra logic at the end of each hint; and between the two I definitely prefer repeating code in derivation.
- I don't think we need to pass subtype readers or writers to hints, in general - once more, I see this as a responsibility of derivation code, giving hints only a way to transform inputs/outputs of reading/writing. Note that if some users really need to call different readers/writers, it seems much more natural to provide custom readers/writers for that purpose.

It seems that a second motivation to change this, besides moving code from derivations to hints, might have been to get rid of that weird `tryNextOnFail` function. I like that idea, but that can be accomplished more nicely by replacing the `Option[ConfigCursor]` in the return type with a richer type that can encode `Found(cur)`, `Tentative(cur)` and `Continue`.

ruippeixotog commented on modules/generic-base/src/main/scala/pureconfig/generic/ProductHint.scala:
```diff
@@ -10,30 +12,80 @@ import pureconfig._
+   *         omitted from the `ConfigObject` representation.
    */
-  def allowUnknownKeys: Boolean
+  def to[A](writer: ConfigWriter[A], fieldName: String, value: A): Option[(String, ConfigValue)]
```
Ditto: consider replacing `writer` and `value` by an `Option[ConfigValue]`.

ruippeixotog commented on modules/generic-base/src/main/scala/pureconfig/generic/ProductHint.scala:
```diff
@@ -10,30 +12,80 @@ import pureconfig._
+   * @return a tuple with both the result of reading the field and the cursor to supply to subsequent reads.
    */
-  def configKey(fieldName: String): String
+  def from[A](cur: ConfigObjectCursor, reader: ConfigReader[A], fieldName: String, default: Option[A]): (ConfigReader.Result[A], ConfigObjectCursor)
```
As an alternative way of defining this, consider not providing `reader` and `default` as arguments and instead providing a richer return type (e.g. `Use(ConfigCursor)`, `UseDefault`). 

This is a good example of providing good flexibility without giving away control flow: if we were to do that here we would have included a `rest: (A, ConfigObjectCursor) => ConfigReader.Result[A]` to keep it consistent with `CoproductHint`).

ruippeixotog commented on modules/generic-base/src/main/scala/pureconfig/generic/ProductHint.scala:
```diff
@@ -10,30 +12,80 @@ import pureconfig._
 
   /**
-   * A boolean indicating if the default arguments of the case class should be used when fields are missing
+   * Returns optional failures if the provided `ConfigObjectCursor` resulted from reading all fields necessary to
```
Did you mean "Returns optional failures given the `ConfigObjectCursor`"? The sentence seems weird as it is right now.

jcazevedo replied:
Yeah. That was weird. Improved it in 8f5bf8d.

ruippeixotog commented on modules/magnolia/src/main/scala/pureconfig/module/magnolia/MagnoliaConfigReader.scala:
```diff
@@ -17,37 +18,21 @@ object MagnoliaConfigReader {
-          if (unknownKeyFailures.isEmpty) res
-          else Left(ConfigReaderFailures(unknownKeyFailures.head, unknownKeyFailures.tail))
-        }
+        res.right.flatMap(values => hint.bottom(obj).fold[ConfigReader.Result[A]](Right(ctx.rawConstruct(values.result())))(Left.apply))
```
If `res` is a `Left` and `hint.bottom(obj)` is `Some`, we should return all failures and not just the ones from `res`. That's the current behavior of `pureconfig-generic` and it seems to be the one that makes more sense.

I see that the previous implementation also has this bug; can you add a test case for this?

jcazevedo replied:
I fIxed that in `MagnoliaConfigReader` and also add a test case in 76d35fc.

ruippeixotog commented on modules/generic/src/main/scala/pureconfig/generic/MapShapedReader.scala:
```diff
@@ -11,7 +10,7 @@ import shapeless.labelled.{ FieldType, field }
  * @tparam Repr the generic representation
  */
-trait MapShapedReader[Wrapped, Repr] extends ConfigReader[Repr]
+trait MapShapedReader[Wrapped, Repr] extends ConfigReader[Wrapped]
```
Why did you change this parameter? `MapShapedReader` is supposed to be a `ConfigReader[Repr]`; `Wrapped` is here just to type the product hint.

jcazevedo replied:
This change was partly motivated by the need to provide `ConfigReader`s of the `Wrapped` type to `CoproductHint`, but this is not relevant anymore, given the current proposal.

ruippeixotog reviewed (changes requested):
Overall looks good @jcazevedo, thanks for doing this! I really like the overhaul of `ProductHint`, it puts it on par with `CoproductHint` while keeping a simplified implementation for users to use.

I have some concerns with generalizing excessively an API that we provide for users to extend (uncommon as it may be). Some of the changes in hints seem a bit confusing and give it too much power to change how the derivation is made - to a point where for certain implementations it may not be recognizable as a derivation anymore. I understood the intention to avoid repeating code between different derivation libs and it worked brilliantly in products, but IMO we can leave control flow in derivation libs (e.g. by avoiding things like `rest` in `CoproductHint`). I left some specific comments inline.

If you did some of these choices to enable new use cases and not just for removing duplicated code, can you provide some interesting examples of how the new capabilities could be used?

ruippeixotog commented:
> This also modifies the derivation of readers for case classes in pureconfig-magnolia to accumulate errors, at the expense of using the unsafe rawConstruct. I ended up doing so because the constructMonadic interface of Magnolia doesn't play very well with the proposed ProductHint interface of returning both the result of reading a field and the next ConfigObjectCursor to use.

This worries me a bit, because it seems that it won't work with the upcoming error accumulating constructor for Magnolia (https://github.com/propensive/magnolia/pull/184) - the current design forces derivations to either follow a recursive approach or use mutable variables. I'm ok with staying with `rawConstruct` if we don't find any good alternative, but if we can think of one we should probably go for it.

jcazevedo committed:
Merge branch 'master' into product-hint-review

jcazevedo committed:
Avoid having to supply case class field values to ProductHint

jcazevedo committed:
Avoid having to supply values to CoproductHints

jcazevedo committed:
Make CoproductHints throw exceptions

jcazevedo committed:
Use `constructEither in MagnoliaConfigReader

jcazevedo committed:
Add test to show that all errors are accumulated with ProductHints that don't allow unknown keys

jcazevedo committed:
Improve wording of ProductHint's `bottom` Scaladoc

jcazevedo committed:
Make MapShapedReader and MapShapedWriter package-private

jcazevedo commented:
@ruippeixotog: Thanks for your thorough review, and sorry for taking so long to address your comments.

I revamped both `ProductHint`s and `CoproductHint`s from my initial proposal, following your suggestion of keeping the control flow in derivation code. The `from` method of `ProductHint` and `CoproductHint` now returns a `FieldHint` and a `ChoiceHint`, respectively (I'm not completely sold on these names). These hints signal intent on what derivation code should do.

> If you did some of these choices to enable new use cases and not just for removing duplicated code, can you provide some interesting examples of how the new capabilities could be used?

Part of the motivation for the `rest` parameter in `CoproductHint` was to allow access to all failures when reading a sealed family (in order to tackle #356). I changed some things in my latest proposal, while also enabling that use case. The `bottom` method of `CoproductHint` now receives a list of all attempted reads and their corresponding failures, allowing better error messages for derived coproduct readers. In order to support that, the `MapShapedReader` implementation changed so that attempts in coproduct reads are accumulated until we reach the bottom type. I suppose it's OK to do that given the "private" nature of `MapShapedReader` and its isolated use in derivation code.

jcazevedo committed:
Avoid links to `ConfigCursor` in Scaladoc

jcazevedo commented:
> This worries me a bit, because it seems that it won't work with the upcoming error accumulating constructor for Magnolia (propensive/magnolia#184) - the current design forces derivations to either follow a recursive approach or use mutable variables. I'm ok with staying with `rawConstruct` if we don't find any good alternative, but if we can think of one we should probably go for it.

With the latest changes, I was able to drop `rawConstruct` and start using `constructEither` from Magnolia. The behavior is slightly different from the shapeless implementation, since keys are only removed from the `ConfigObject` after all fields are read, whereas in the shapeless implementation they are removed after reading each field. I think both implementations should have the same end result unless the provided field mapping is non-injective, in which case other types of problems should show up first.

ruippeixotog committed:
Merge branch 'master' into product-hint-review

ruippeixotog commented on modules/generic-base/src/main/scala/pureconfig/generic/CoproductHint.scala:
```diff
@@ -14,49 +15,38 @@ import pureconfig.syntax._
    */
-  def noOptionFound(cur: ConfigCursor): ConfigReaderFailures =
-    ConfigReaderFailures(cur.failureFor(NoValidCoproductChoiceFound(cur.value)))
+  def to(name: String, cv: ConfigValue): ConfigValue
```
Can we switch the two arguments back to their original positions? This would make it consistent with `from` and result in an easier migration.

ruippeixotog commented on modules/generic-base/src/main/scala/pureconfig/generic/CoproductHint.scala:
```diff
@@ -78,34 +68,35 @@ class FieldCoproductHint[T](key: String) extends CoproductHint[T] {
+  override def bottom(cur: ConfigCursor, attempts: List[(String, ConfigReaderFailures)]): ConfigReaderFailures =
+    attempts match {
+      case h :: _ => h._2
+      case _ => (for {
```
This structure is rather hacky; we are relying on the implementation of `from` in that we know that 1) the only way no attempts were made was if there's something wrong with the key, 2) we'll only try at most one key.

I think this is not really a case for `bottom`, but because `from` does not return an `Either` anymore we have no way of saying "an error occurred" (when I suggested moving to a richer type I did not mean getting rid of the `Result` part). What do you think of allowing `from` calls to signal an error and stopping the process? That can be either by wrapping in a `Result` (my preference, for consistency with the rest of the lib) or adding a 4th class `Error` to `ChoiceHint`.

ruippeixotog commented on modules/generic-base/src/main/scala/pureconfig/generic/CoproductHint.scala:
```diff
@@ -78,34 +68,35 @@ class FieldCoproductHint[T](key: String) extends CoproductHint[T] {
+        objCur <- cur.asObjectCursor.right
+        valueCur <- objCur.atKey(key).right
+        _ <- valueCur.asString.right
+      } yield ConfigReaderFailures(valueCur.failureFor(UnexpectedValueForFieldCoproductHint(valueCur.value)))).fold(identity, identity)
```
Nit: it's really confusing to understand for comprehensions wrapped in parentheses with methods chained afterwards, as it makes for huge lines like this one and indentation is not well handled. When that happens, consider refactoring the code or using auxiliary variables.

ruippeixotog commented on modules/generic-base/src/main/scala/pureconfig/generic/CoproductHint.scala:
```diff
@@ -78,34 +68,35 @@ class FieldCoproductHint[T](key: String) extends CoproductHint[T] {
+  def to(name: String, cv: ConfigValue): ConfigValue = {
+    cv match {
+      case co: ConfigObject =>
+        if (co.containsKey(key))
```
Nit: consider merging this `if` into the pattern match above for better readability.

ruippeixotog commented on modules/generic-base/src/main/scala/pureconfig/generic/CoproductHint.scala:
```diff
@@ -117,11 +108,42 @@ object FieldCoproductHint {
+  /**
+   * A hint on what should be done when reading a given coproduct option.
+   */
+  sealed trait ChoiceHint
```
May I suggest `Action` or `Decision` instead of anything involving `Hint`? This has very little in common with `ProductHint`s and `CoproductHint`s, which may be confusing to users.

ruippeixotog commented on modules/generic-base/src/main/scala/pureconfig/generic/ProductHint.scala:
```diff
@@ -10,34 +13,73 @@ import pureconfig._
+   *         omitted from the `ConfigObject` representation.
    */
-  def allowUnknownKeys: Boolean
+  def to(fieldName: String, value: Option[ConfigValue]): Option[(String, ConfigValue)]
```
Same thing here wrt swapping args.

ruippeixotog commented on modules/generic-base/src/main/scala/pureconfig/generic/ProductHint.scala:
```diff
@@ -10,34 +13,73 @@ import pureconfig._
+   * @param remove whether the field should be removed from the object after read
+   * @param useDefault whether to use default values when reading this field if the cursor is undefined
+   */
+  case class FieldHint(cursor: ConfigCursor, field: String, remove: Boolean, useDefault: Boolean)
```
Ditto here about the name. `Action`/`Decision` still sound good to me here.

It seems strange to pass all this metadata together with the real purpose of the hint (the `cursor` to be used to read the key). I wonder if we should rethink the assumption that all field mappings are non-injective and that we have to _remove_ the config keys as they are read - this makes things a bit confusing any not parallelizable (in the sense that we cannot use the power of `Applicative` to process fields independently and join errors while keeping the same behavior).

Semantically, `!allowUnknownKeys` only requires that at the end no keys were unused. Can't we just have this `FieldHint` return a list of `usedFields` (instead of `field`+`remove`), which are either passed to `bottom` or used to filter the config object before calling `bottom`? This paves the way for hints that read fields multiple times (e.g. case classes having multiple fields with different representations of the same config value) and hints that read several config keys for a single field (e.g. reading a number and a unit to form a [coulomb](https://github.com/erikerlandson/coulomb) dimension).

Regarding `useDefault`: just like we have different actions like `Attempt` and `Skip`, I wonder if we shouldn't have a `UseDefault` or `UseOrDefault(ConfigCursor)` here for consistency? This is more a style concern, I don't see any immediate advantage.

ruippeixotog commented on modules/generic/src/main/scala/pureconfig/generic/MapShapedReader.scala:
```diff
@@ -71,30 +67,34 @@ object MapShapedReader {
+              .right.map(v => Inl(field[Name](v)))
+              .left.map(failures => coproductHint.bottom(cur, attempts :+ (vName.value.name -> failures)))
+          case Attempt(optCur) =>
+            vReader.from(optCur)
```
Can you please extract this to an auxiliary function?

ruippeixotog commented on modules/magnolia/src/main/scala/pureconfig/module/magnolia/MagnoliaConfigReader.scala:
```diff
@@ -17,37 +18,44 @@ object MagnoliaConfigReader {
+        val res = ctx.constructEither[ConfigReaderFailures, Param[ConfigReader, A]#PType] { param =>
+          val fieldHint = hints(param.label)
+          lazy val reader = param.typeclass
+          lazy val keyNotFoundFailure = cur.failed(KeyNotFound.forKeys(fieldHint.field, objCur.keys))
```
Why are we creating this here instead of creating it inline on `case _ =>`?

ruippeixotog commented on modules/magnolia/src/main/scala/pureconfig/module/magnolia/MagnoliaConfigReader.scala:
```diff
@@ -17,37 +18,44 @@ object MagnoliaConfigReader {
+            currObj
         }
+
+        hint.bottom(filteredObj).fold(
```
This can be greatly simplified to `ConfigReader.Result.zipWith(res, hint.bottom(filteredObj).toLeft(()))((r, _) => r)`.

ruippeixotog commented on modules/magnolia/src/main/scala/pureconfig/module/magnolia/MagnoliaConfigReader.scala:
```diff
@@ -17,37 +18,44 @@ object MagnoliaConfigReader {
-          else Left(ConfigReaderFailures(unknownKeyFailures.head, unknownKeyFailures.tail))
+        val filteredObj = hints.foldLeft(objCur) {
+          case (currObj, (_, FieldHint(_, field, true, _))) =>
+            currObj.withoutKey(field)
```
Now that we don't need any of the intermediate results and in most cases this config object will end up empty, it's more efficient and cleaner if we collect all used keys and create a single new config object with all of them filtered out. We can create such a method on `ConfigObject` to support this (probably `filterKeys`).

ruippeixotog commented on modules/magnolia/src/main/scala/pureconfig/module/magnolia/MagnoliaConfigReader.scala:
```diff
@@ -67,28 +75,33 @@ object MagnoliaConfigReader {
+    def from(cur: ConfigCursor): ConfigReader.Result[A] = {
+      val (_, res) = ctx.subtypes.foldLeft[(Boolean, Either[List[(String, ConfigReaderFailures)], A])]((true, Left(Nil))) {
+        case (acc, subtype) =>
+          acc match {
```
Is there any reason why you split this pattern match in two nested ones? I find it easier to read if you have a single matching level, and if you put `case ((true, Left(attempts)), subtype) =>` as your first case you only have `case (acc, _) => acc` as your second one, just like we had before.

ruippeixotog commented on modules/magnolia/src/main/scala/pureconfig/module/magnolia/MagnoliaConfigReader.scala:
```diff
@@ -67,28 +75,33 @@ object MagnoliaConfigReader {
+              val typeName = subtype.typeName.short
+              hint.from(cur, typeName) match {
+                case Use(cur) =>
+                  subtype.typeclass.from(cur).fold(
```
Ditto here regarding moving to an aux function.

ruippeixotog commented on modules/magnolia/src/main/scala/pureconfig/module/magnolia/MagnoliaConfigWriter.scala:
```diff
@@ -22,12 +21,13 @@ object MagnoliaConfigWriter {
-          case tc =>
-            Some(hint.configKey(param.label) -> tc.to(param.dereference(a)))
+        val valueOpt = param.typeclass match {
+          case w: WritesMissingKeys[param.PType @unchecked] =>
```
Any reason why you renamed this? (just out of curiosity)

ruippeixotog commented on modules/generic/src/main/scala/pureconfig/generic/MapShapedReader.scala:
```diff
@@ -11,7 +13,9 @@ import shapeless.labelled.{ FieldType, field }
  * @tparam Repr the generic representation
  */
-trait MapShapedReader[Wrapped, Repr] extends ConfigReader[Repr]
+private[generic] trait MapShapedReader[Wrapped, Repr] {
```
If we are making this private, we should be consistent and make all other related ones, too (`Seq` and `Map` shaped readers and writers).

ruippeixotog commented on modules/generic/src/main/scala/pureconfig/generic/MapShapedReader.scala:
```diff
@@ -48,21 +46,19 @@ object MapShapedReader {
-        case keyCur => headReader.from(keyCur)
+      val fieldHint = hint.from(cur, fieldName)
+      lazy val reader = vFieldReader.value.value
+      lazy val keyNotFoundFailure = cur.failed[V](KeyNotFound.forKeys(fieldHint.field, cur.keys))
```
Why are we creating this here instead of creating it inline on `case _ =>`?

ruippeixotog commented on modules/generic-base/src/main/scala/pureconfig/generic/error/CoproductHintException.scala:
```diff
@@ -0,0 +1,7 @@
+
+import pureconfig.error.FailureReason
+
+final case class CoproductHintException(failure: FailureReason) extends RuntimeException {
```
Can you add Scaladoc to this?

jcazevedo committed:
Reorder arguments' position

jcazevedo committed:
Allow the `from` method of `CoproductHint` to return an `Either`

jcazevedo committed:
Rename `CoproductHint.ChoiceHint` to `CoproductHint.Action`

jcazevedo committed:
Rename `ProductHint.FieldHint` to `ProductHint.Action`

jcazevedo committed:
Rename `cur` to `cursor`

jcazevedo committed:
Signal to use default value directly at action type and collect used fields for bottom

jcazevedo committed:
Only create failure on relevant case option

jcazevedo committed:
Add auxiliary functions to `cConsReader` implementation

jcazevedo committed:
Simplify accumulation of failures

jcazevedo committed:
Use a single level of pattern matching

jcazevedo committed:
Use an auxiliary function for attempting reads

jcazevedo committed:
Rename typeclass variable

jcazevedo committed:
Make helper traits of the `generic` package private

jcazevedo committed:
Add Scaladoc to `CoproductHintException`

jcazevedo committed:
Don't make too much stuff private

ruippeixotog committed:
Merge branch 'master' into product-hint-review

jcazevedo committed:
Have Use and Attempt as the only available CoproductHint options

jcazevedo committed:
Don't use ConfigReader.Result.succeed

jcazevedo committed:
Merge branch 'master' into product-hint-review

jcazevedo committed:
Don't link to ConfigReader.Result

jcazevedo committed:
Use right projection for Scala 2.11

jcazevedo committed:
Break long lines

jcazevedo committed:
Improve performance of derived readers

jcazevedo committed:
Avoid using MapShapedWriter for coproduct derivation

jcazevedo committed:
Make options in CoproductReaderOptions lazy

jcazevedo committed:
Merge branch 'master' into product-hint-review

jcazevedo commented:
@ruippeixotog: Sorry for taking so long to get back to this. I think I addressed all your comments, but I ended up changing some other things I wasn't very happy with.

I wasn't very happy with how the actions of `CoproductHint` were dependent on the order the coproduct options were visited. Even though both Shapeless and Magnolia guarantee a stable ordering of coproduct options, we have no guarantees that the order will be the same between them. To take this into account, I modified `CoproductHint`'s `from` method to receive the names of the coproduct options. Implementations of this method can then choose to use one of the options (with `CoproductHint.Use`) or attempt various options, in a defined order (with `Coproduct.Attempt`). 

To implement this, I introduced the `CoproductReaderOptions` type class to collect `ConfigReader`s for all coproduct options of a given type, indexed by their name. This also stops considering `ConfigReader`s for coproducts as `MapShapedReader`s, which is also more accurate. In order to be consistent in the writing side, I introduced the `CoproductConfigWriter` type class, so that `ConfigWriter`s for coproducts stop being `MapShapedWriter`s.

jcazevedo committed:
Merge branch 'master' into product-hint-review

ruippeixotog commented on modules/generic-base/src/main/scala/pureconfig/generic/ProductHint.scala:
```diff
@@ -10,34 +12,101 @@ import pureconfig._
+      case (k, keyCur) if !usedFields.contains(k) =>
+        keyCur.failureFor(UnknownKey(k))
+    }
+    if (allowUnknownKeys || unknownKeys.isEmpty)
```
nit: can we use pattern matching on `unknownKeys` instead of `.isEmpty`/`.head`/`.tail`? I'd prefer to avoid these and lazy vals, which make the flow fuzzier.

jcazevedo replied:
Sure. Improved that in a0b3fb8d.

ruippeixotog commented on modules/generic-base/src/main/scala/pureconfig/generic/ProductHint.scala:
```diff
@@ -10,34 +12,101 @@ import pureconfig._
+  case class Use(cursor: ConfigCursor, field: String) extends Action
+
+  /**
+   * An action to either use the provided `ConfigCursor` (if it isn't null nor undefined) or fallback to the default
```
```suggestion
   * An action to either use the provided `ConfigCursor` (if it isn't null or undefined) or fallback to the default
```

jcazevedo replied:
Fixed in 34f3a4bb.

ruippeixotog commented on modules/generic-base/src/main/scala/pureconfig/generic/error/MissingCoproductChoice.scala:
```diff
@@ -0,0 +1,12 @@
+ *
+ * @param option the coproduct option that is missing
+ */
+final case class MissingCoproductChoice(option: String) extends FailureReason {
```
nit: consider renaming this `InvalidCoproductOption` or `InvalidCoproductOptionChosen` - we should be consistent with the term "coproduct option" to refer to subtypes here and I prefer "invalid" because it makes it clear that the chosen option is not supposed to be there in the first place (i.e. that this is a bug on the hint implementation).

jcazevedo replied:
Make sense. I've started using "coproduct option" instead of "coproduct choice" throughout the codebase in f5013e48.

ruippeixotog commented on modules/generic-base/src/main/scala/pureconfig/generic/error/MissingCoproductChoice.scala:
```diff
@@ -0,0 +1,12 @@
+ * @param option the coproduct option that is missing
+ */
+final case class MissingCoproductChoice(option: String) extends FailureReason {
+  def description = s"Missing coproduct choice '$option''"
```
I'd also add here that this is likely to be a bug on the coproduct hint implementation, as the only way this can happen is if users violate the contract of `CoproductHint#from`. As it is now the description is too broad.

jcazevedo replied:
Made the description more specific in f5013e48.

ruippeixotog commented on modules/generic/src/main/scala/pureconfig/generic/EnumCoproductHint.scala:
```diff
@@ -1,8 +1,10 @@
 import pureconfig._
 import pureconfig.error._
+import pureconfig.generic.CoproductHint.Use
+import pureconfig.generic.error.{ CoproductHintException, NoValidCoproductChoiceFound }
```
I noticed now that we already had a `NoValidCoproductChoiceFound` failure. We can take this opportunity to rename it to "option", if you want.

ruippeixotog commented on modules/generic/src/main/scala/pureconfig/generic/EnumCoproductHint.scala:
```diff
@@ -22,18 +24,20 @@ class EnumCoproductHint[T] extends CoproductHint[T] {
+    value match {
+      case co: ConfigObject if co.isEmpty => fieldValue(name).toConfig
+      case _: ConfigObject =>
+        throw new CoproductHintException(NonEmptyObjectFound(name))
```
```suggestion
        throw CoproductHintException(NonEmptyObjectFound(name))
```

jcazevedo replied:
Removed the unnecessary `new`s in a1c7fe6c.

ruippeixotog commented on modules/generic/src/main/scala/pureconfig/generic/EnumCoproductHint.scala:
```diff
@@ -22,18 +24,20 @@ class EnumCoproductHint[T] extends CoproductHint[T] {
+      case _: ConfigObject =>
+        throw new CoproductHintException(NonEmptyObjectFound(name))
+      case cv =>
+        throw new CoproductHintException(WrongType(cv.valueType, Set(ConfigValueType.OBJECT)))
```
```suggestion
        throw CoproductHintException(WrongType(cv.valueType, Set(ConfigValueType.OBJECT)))
```

ruippeixotog commented on modules/generic/src/main/scala/pureconfig/generic/MapShapedReader.scala:
```diff
@@ -1,100 +1,55 @@
+        val fieldName = key.value.name
+        val fieldAction = hint.from(cur, fieldName)
+        lazy val reader = vConfigReader.value.value
+        lazy val keyNotFoundFailure = cur.failed[V](KeyNotFound.forKeys(fieldAction.field, cur.keys))
```
Can we move this to the only `case` where it's used? It's generally harder to read lazy vals with no context of where they're going to be used.

jcazevedo replied:
Improved in 2fb7f84e.

ruippeixotog commented on modules/generic/src/main/scala/pureconfig/generic/DerivedConfigReader.scala:
```diff
@@ -64,20 +66,42 @@ trait DerivedConfigReader1 {
     }
   }
 
   final implicit def coproductReader[F, Repr <: Coproduct](
```
I'm not really sure why we chose `F` for this type parameter, but it's a strange choice as `F` is typically used for higher-kinded types...

jcazevedo replied:
Not really sure. I'd rather fix that in a different PR, though.

ruippeixotog commented on modules/generic/src/main/scala/pureconfig/generic/DerivedConfigReader.scala:
```diff
@@ -64,20 +66,42 @@ trait DerivedConfigReader1 {
+            curr.left.flatMap { currentFailures =>
+              readerFor(option) match {
+                case Some(value) => value.from(cursor).left.map(f => currentFailures :+ (option -> f))
+                case None => Left(currentFailures :+ (option -> ConfigReaderFailures(cursor.failureFor(MissingCoproductChoice(option)))))
```
nit: please add a newline somewhere in this line.

jcazevedo replied:
Added newline in f5013e48.

ruippeixotog commented on modules/magnolia/src/main/scala/pureconfig/module/magnolia/MagnoliaConfigReader.scala:
```diff
@@ -67,27 +62,33 @@ object MagnoliaConfigReader {
+            curr.left.flatMap { currentFailures =>
+              readerFor(option) match {
+                case Some(value) => value.from(cur).left.map(f => currentFailures :+ (option -> f))
+                case None => Left(currentFailures :+ (option -> ConfigReaderFailures(cur.failureFor(MissingCoproductChoice(option)))))
```
nit: newline here too

jcazevedo replied:
Added newline in f5013e48.

ruippeixotog reviewed (approved):
LGTM overall! I left a few minor comments around, please take a look.

I'm really looking forward to merge this, after almost 7 months of backs and forths :)

jcazevedo committed:
Merge branch 'master' into product-hint-review

jcazevedo committed:
Pattern match on `unknownKeys`

jcazevedo committed:
Fix wording in docs

jcazevedo committed:
Prefer coproduct option to coproduct choice

jcazevedo committed:
Remove unnecessary `new`s

jcazevedo committed:
Avoid unnecessary lazy val

ruippeixotog commented:
Travis is failing because html-proofer is failing to get bootstrap.min.css. I'll merge this anyway as it's unrelated to this PR. Thanks again for this refactoring, @jcazevedo!
